### PR TITLE
Initial `Call` implementation

### PR DESCRIPTION
**Pull request by ascjones**
Extracted from https://github.com/paritytech/ink/pull/124.

Provides a basic `Call` implementation for use with `ext_dispatch_call`, for calling substrate modules from contracts.

This PR provides only the `Balances` variant as a proof of concept, together with an example contract which dispatches a `transfer` call.

Further modules/call variants will be added in future PRs.

It is expected that chain and module authors will construct their own `Call` enum types compatible with their custom runtime. They will likely copy the types here as a basis for building those.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
 /target
 **/*.rs.bk
 Cargo.lock
+.idea
+rust-toolchain
```
Comment (Robbepop) [inline .gitignore]: We should no longer use `rust-toolchain` files.
Did you include this to protect against including a `rust-toolchain` file ever again? :D

File: Cargo.toml [hunk 2]
```diff
@@ -18,18 +18,20 @@ categories = ["no-std", "embedded"]
 include = ["/Cargo.toml", "src/**/*.rs", "/README.md", "/LICENSE"]
 
 [dependencies]
-# FIXME: remove aj-runtime-types branch once https://github.com/paritytech/ink/pull/108 merged
-ink_core = { git = "https://github.com/paritytech/ink/", branch = "aj-runtime-types", package = "ink_core" }
-parity-codec = { version = "3.2", default-features = false, features = ["derive"] }
+ink_core = { git = "https://github.com/paritytech/ink/", branch = "aj-dispatch-call", package = "ink_core", default-features = false }
```
Comment (Robbepop) [inline Cargo.toml]: We want to merge your ink! PR before this so we can adjust lines like these before merging this PR.

File: examples/calls/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,43 @@
+[package]
+name = "calls"
+version = "0.1.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
+
+[dependencies]
+ink_core = { git = "https://github.com/paritytech/ink/", branch = "aj-dispatch-call", package = "ink_core", default-features = false }
```
Comment (Robbepop) [inline examples/calls/Cargo.toml]: Can be changed to `master` branch now.

File: examples/calls/.cargo/config [hunk 4]
```diff
@@ -0,0 +1,5 @@
+[target.wasm32-unknown-unknown]
+rustflags = [
+	"-C", "overflow-checks=on",
```
Comment (Robbepop) [inline examples/calls/.cargo/config]: https://github.com/paritytech/ink/pull/158

File: examples/calls/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    "ink_lang/test-env",
+]
+generate-api-description = [
+    "ink_lang/generate-api-description"
+]
+
+[profile.release]
+panic = "abort"
+lto = true
+opt-level = "z"
```
Comment (Robbepop) [inline examples/calls/Cargo.toml:43]: https://github.com/paritytech/ink/pull/158

File: src/calls.rs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+use core::convert::TryInto;
+
+#[cfg_attr(feature = "std", derive(Debug, Clone, PartialEq, Eq))]
+pub enum Address<T: EnvTypes, AccountIndex> {
+    Id(T::AccountId),
+    Index(AccountIndex),
+}
+
+/// Returns `b` if `b` is greater than `a` and otherwise `None`.
+fn need_more_than<T: PartialOrd>(a: T, b: T) -> Option<T> {
```
Comment (Robbepop) [inline src/calls.rs]: Please rename this.

File: .gitignore (hunk 1 above)
Comment (ascjones) [reply .gitignore]: I'm using a `rust-toolchain` file locally so I don't need to always add `+nightly` to my cargo commands. I'm not cool enough to have nightly as my default toolchain :grin:

File: src/calls.rs (hunk 6 above)
Comment (ascjones) [reply src/calls.rs]: Copied from substrate, but couldn't think of a better name. Have gone for `greater_than_or_none`

File: .gitignore (hunk 1 above)
Comment (Robbepop) [reply .gitignore]: You could try ignore it locally instead via the `.git/info/exclude` file that has the same format as `.gitignore` but acts locally. Same for `.idea` ignore.

Comment (ascjones) [reply .gitignore]: Wow didn't know about this. Thanks!

Comment (Robbepop) [reply .gitignore]: As you see I didn't know either at the time when I put VS Code stuff into the `.gitignore`. :D

**Pull request merged**